map_model/src/pathfind/v2.rs

```diff
@@ -3,82 +3,181 @@
-            let base = 1;
-            base + extra_penalty
-        },
-        |_| 0,
-    ) {
-        Some((_, path)) => {
-            let mut steps = Vec::new();
-            for pair in path.windows(2) {
-                steps.push(PathStep::Lane(pair[0]));
-                // We don't need to look for this turn in the map; we know it exists.
-                steps.push(PathStep::Turn(TurnID {
-                    parent: map.get_l(pair[0]).dst_i,
-                    src: pair[0],
-                    dst: pair[1],
-                }));
+        }
+
+        match petgraph::algo::astar(
+            &graph,
+            self.req.start.lane(),
+            |l| l == self.req.end.lane(),
+            |(_, _, t)| {
+                // Normally opportunistic lane-changing adjusts the path live, but that doesn't work
+                // near uber-turns. So still use some of the penalties here.
+                let (lt, lc, slow_lane) = map.get_t(*t).penalty(map);
+                let mut extra_penalty = lt + lc;
+                if self.req.constraints == PathConstraints::Bike {
+                    extra_penalty = slow_lane;
+                }
+                // Always treat every lane/turn as at least cost 1; otherwise A* can't understand
+                // that a final path with 10 steps costs more than one with 5. The
+                // road-based pathfinding has already chosen the overall route; when
+                // we're picking individual lanes, the length of each lane along one
+                // road is going to be about the same.
+                let base = 1;
+                base + extra_penalty
+            },
+            |_| 0,
+        ) {
+            Some((_, path)) => {
+                let mut steps = Vec::new();
+                for pair in path.windows(2) {
+                    steps.push(PathStep::Lane(pair[0]));
+                    // We don't need to look for this turn in the map; we know it exists.
+                    steps.push(PathStep::Turn(TurnID {
+                        parent: map.get_l(pair[0]).dst_i,
+                        src: pair[0],
+                        dst: pair[1],
+                    }));
+                }
+                steps.push(PathStep::Lane(self.req.end.lane()));
+                assert_eq!(steps[0], PathStep::Lane(self.req.start.lane()));
+                let uber_turns = find_uber_turns(&steps, map, self.uber_turns);
+                Ok(Path::new(map, steps, self.req, uber_turns))
             }
-            steps.push(PathStep::Lane(req.end.lane()));
-            assert_eq!(steps[0], PathStep::Lane(req.start.lane()));
-            let uber_turns = find_uber_turns(&steps, map, uber_turns_v2);
-            Ok(Path::new(map, steps, req, uber_turns))
+            None => bail!(
+                "Can't transform a road-based path to a lane-based path for {}",
+                self.req
+            ),
+        }
+    }
+
+    fn to_v1_walking(self, map: &Map) -> Result<Path> {
```

Comment:
This is happily straightforward, because the correspondence between a `DirectedRoadID` and the walkable lane on it is 1:1 when that directed road is walkable at all.